src/middlewares/__tests__/user-authentication.test.js

```diff
@@ -0,0 +1,148 @@
+const jwt = require('jsonwebtoken');
+const userAuthentication = require('../user-authentication');
+
+const validTokenWithValidTime = {
+  tokenValue: 'validToken',
+  expires: 'validTime',
+};
+const validTokenWithNoValidTime = {
+  tokenValue: 'validToken',
+  expires: 'invalidTime',
+};
+const invalidTokenWithValidTime = {
+  tokenValue: 'invalidToken',
+  expires: 'validTime',
+};
+const invalidTokenWithInvalidTime = {
+  tokenValue: 'invalidToken',
+  expires: 'invalidTime',
+};
+const mockUser = {
+  id: 1,
+  username: 'name',
+  email: 'someemail@gmail.com',
+};
+
+const res = {
+  status: jest.fn().mockReturnThis(),
+  json: jest.fn().mockReturnThis(),
+  clearCookie: jest.fn().mockReturnThis(),
+};
+const next = jest.fn();
+
+const verify = jest.spyOn(jwt, 'verify');
+
+afterAll(() => {
+  jest.clearAllMocks();
+});
+describe('user-authentication function middleware', () => {
+  it('Should authenticate for valid token with valid expiry time', () => {
+    verify.mockImplementation(() => ({ user: mockUser }));
+    const req = {
+      cookies: {
+        token: validTokenWithValidTime,
+      },
+    };
+
+    userAuthentication(req, res, next);
+    expect(verify).toHaveReturnedWith({ user: mockUser });
+    expect(verify).toHaveBeenCalled();
+    expect(verify).toHaveBeenCalledTimes(1);
+    expect(res.json).not.toHaveBeenCalled();
+    expect(verify).not.toThrow();
+    expect(verify).toHaveBeenCalledWith(
+      validTokenWithValidTime,
+      process.env.JWT_SECRET
+    );
+    expect(next).toHaveBeenCalled();
+    expect(req.user).toEqual(mockUser);
+    expect(req.user.id).toEqual(mockUser.id);
```

Comment:
Here as well try to keep the order of execution in the order of your expect statements.

So first check `req.user`, then do all the checks on `verify` and last on `next` and `res.json` or `res.status`.

Also always run `toHaveBeenCalled` first, then `toHaveBeenCalledTimes`, then `toHaveBeenCalledWith` and last `toHaveReturnedWith` or `toThrow`.